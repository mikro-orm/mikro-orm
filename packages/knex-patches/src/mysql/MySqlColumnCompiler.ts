// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true }) {
    return 'int unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true }) {
    return 'bigint unsigned not null auto_increment' + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
