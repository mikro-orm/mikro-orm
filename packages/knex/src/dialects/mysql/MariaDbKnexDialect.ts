import { MonkeyPatchable } from '../../MonkeyPatchable';

export class MariaDbKnexDialect extends MonkeyPatchable.MySqlDialect {

  get driverName() {
    return 'mariadb';
  }

  _driver() {
    return require('mariadb/callback');
  }

  validateConnection(connection: any) {
    return connection.isValid();
  }

}
