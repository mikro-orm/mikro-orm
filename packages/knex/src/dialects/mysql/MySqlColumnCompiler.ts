// @ts-ignore
import BaseMySqlColumnCompiler from 'knex/lib/dialects/mysql/schema/mysql-columncompiler';

export class MySqlColumnCompiler extends BaseMySqlColumnCompiler {

  // we need the old behaviour to be able to add auto_increment to a column that is already PK
  increments(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'int unsigned' : 'int';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

  /* istanbul ignore next */
  bigincrements(this: any, options = { primaryKey: true, unsigned: true }) {
    const type = options.unsigned ? 'bigint unsigned' : 'bigint';
    return `${type} not null auto_increment` + (this.tableCompiler._canBeAddPrimaryKey(options) ? ' primary key' : '');
  }

}
