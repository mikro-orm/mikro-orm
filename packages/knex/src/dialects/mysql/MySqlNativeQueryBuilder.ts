import { type Dictionary, LockMode, RawQueryFragment, Utils } from '@mikro-orm/core';
import { NativeQueryBuilder } from '../../query/NativeQueryBuilder.js';

/** @internal */
export class MySqlNativeQueryBuilder extends NativeQueryBuilder {

  protected override compileInsert() {
    if (!this.options.data) {
      throw new Error('No data provided');
    }

    this.parts.push('insert');

    if (this.options.onConflict?.ignore) {
      this.parts.push('ignore');
    }

    this.addHintComment();
    this.parts.push(`into ${this.getTableName()}`);

    if (Object.keys(this.options.data).length === 0) {
      this.parts.push('default values');
      return;
    }

    const parts = this.processInsertData();
    this.parts.push(parts.join(', '));
  }

  protected override addLockClause() {
    if (!this.options.lockMode) {
      return;
    }

    const map = {
      [LockMode.PESSIMISTIC_READ]: 'lock in share mode',
      [LockMode.PESSIMISTIC_WRITE]: 'for update',
      [LockMode.PESSIMISTIC_PARTIAL_WRITE]: 'for update skip locked',
      [LockMode.PESSIMISTIC_WRITE_OR_FAIL]: 'for update nowait',
      [LockMode.PESSIMISTIC_PARTIAL_READ]: 'lock in share mode skip locked',
      [LockMode.PESSIMISTIC_READ_OR_FAIL]: 'lock in share mode nowait',
    } as const;

    if (this.options.lockMode !== LockMode.OPTIMISTIC) {
      this.parts.push(map[this.options.lockMode]);
    }
  }

  protected override addOnConflictClause() {
    const clause = this.options.onConflict;

    if (!clause || clause.ignore) {
      return;
    }

    if (clause.merge) {
      this.parts.push('on duplicate key update');

      if (Utils.isObject(clause.merge)) {
        const fields = Object.keys(clause.merge).map(field => {
          this.params.push((clause.merge as Dictionary)[field]);
          return `${this.quote(field)} = ?`;
        });
        this.parts.push(fields.join(', '));
      } else if ((clause.merge as string[]).length === 0) {
        const dataAsArray = Utils.asArray(this.options.data);
        const keys = Object.keys(dataAsArray[0]);
        this.parts.push(keys.map(key => `${this.quote(key)} = values(${this.quote(key)})`).join(', '));
      } else {
        const fields = (clause.merge as string[]).map(key => `${this.quote(key)} = values(${this.quote(key)})`);
        this.parts.push(fields.join(', '));
      }

      if (clause.where) {
        this.parts.push(`where ${clause.where.sql}`);
        this.params.push(...clause.where.params);
      }

      return;
    }

    this.parts.push('on conflict');

    if (clause.fields instanceof RawQueryFragment) {
      this.parts.push(clause.fields.sql);
      this.params.push(...clause.fields.params);
    } else if (clause.fields.length > 0) {
      const fields = clause.fields.map(field => this.quote(field));
      this.parts.push(`(${fields.join(', ')})`);
    }
  }

}
