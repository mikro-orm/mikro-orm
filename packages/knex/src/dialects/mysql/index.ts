export * from './MariaDbKnexDialect';
export * from './MySqlConnection';
export * from './MySqlExceptionConverter';
export * from './MySqlSchemaHelper';
export * from './MySqlPlatform';
export * from './MySqlNativeQueryBuilder';
