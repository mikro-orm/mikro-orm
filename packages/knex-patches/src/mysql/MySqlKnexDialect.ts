import { MySqlQueryCompiler } from './MySqlQueryCompiler';
import { MySqlColumnCompiler } from './MySqlColumnCompiler';
import { MonkeyPatchable } from '../MonkeyPatchable';

export class MySqlKnexDialect extends MonkeyPatchable.MySqlDialect {

  queryCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlQueryCompiler as any)(this, ...arguments);
  }

  columnCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlColumnCompiler as any)(this, ...arguments);
  }

}
