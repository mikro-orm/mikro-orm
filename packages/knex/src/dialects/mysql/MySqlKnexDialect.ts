// @ts-ignore
import MySql2Dialect from 'knex/lib/dialects/mysql2';
import { MySqlQueryCompiler } from './MySqlQueryCompiler';
import { MySqlColumnCompiler } from './MySqlColumnCompiler';

export class MySqlKnexDialect extends MySql2Dialect {

  queryCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlQueryCompiler as any)(this, ...arguments);
  }

  columnCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (MySqlColumnCompiler as any)(this, ...arguments);
  }

}
