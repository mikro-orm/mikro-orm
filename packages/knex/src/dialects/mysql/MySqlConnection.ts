import type { Knex } from 'knex';
import { MySqlKnexDialect } from './MySqlKnexDialect';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';

export class MySqlConnection extends AbstractSqlConnection {

  override createKnex() {
    this.client = this.createKnexClient(MySqlKnexDialect as any);
    this.connected = true;
  }

  getDefaultClientUrl(): string {
    return 'mysql://root@127.0.0.1:3306';
  }

  override getConnectionOptions(): Knex.MySqlConnectionConfig {
    const ret = super.getConnectionOptions() as Knex.MySqlConnectionConfig;

    if (this.config.get('multipleStatements')) {
      ret.multipleStatements = this.config.get('multipleStatements');
    }

    if (this.config.get('forceUtcTimezone')) {
      ret.timezone = 'Z';
    }

    if (this.config.get('timezone')) {
      ret.timezone = this.config.get('timezone');
    }

    ret.supportBigNumbers = true;
    ret.dateStrings = true;

    return ret;
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'run' && ['OkPacket', 'ResultSetHeader'].includes(res[0].constructor.name)) {
      return {
        insertId: res[0].insertId,
        affectedRows: res[0].affectedRows,
        rows: [],
      } as unknown as T;
    }

    if (method === 'get') {
      return res[0][0];
    }

    return res[0];
  }

}
