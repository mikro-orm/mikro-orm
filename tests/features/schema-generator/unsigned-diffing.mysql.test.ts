import { Entity, MikroORM, PrimaryKey, Property } from '@mikro-orm/mysql';

@Entity({ tableName: 'book' })
class Book0 {

  @PrimaryKey()
  id!: number;

  @Property()
  price1!: number;

  @Property({ type: 'decimal' })
  price2!: string;

  @Property({ type: 'double' })
  price3!: string;

}

@Entity({ tableName: 'book' })
class Book1 {

  @PrimaryKey()
  id!: number;

  @Property({ unsigned: true })
  price1!: number;

  @Property({ type: 'decimal', unsigned: true })
  price2!: string;

  @Property({ type: 'double', unsigned: true })
  price3!: string;

}

@Entity({ tableName: 'book' })
class Book2 {

  @PrimaryKey()
  id!: number;

  @Property()
  price1!: number;

  @Property({ type: 'decimal' })
  price2!: string;

  @Property({ type: 'double' })
  price3!: string;

}

describe('unsigned diffing in mysql', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Book0],
      dbName: `mikro_orm_test_unsigned_diffing`,
      port: 3308,
    });
    await orm.schema.ensureDatabase();
    await orm.schema.execute('drop table if exists book');
    await orm.schema.createSchema();
  });

  afterAll(() => orm.close(true));

  test('schema generator updates column types when length changes (varchar, decimal, ...)', async () => {
    orm.discoverEntity(Book1, 'Book0');
    const diff1 = await orm.schema.getUpdateSchemaMigrationSQL({ wrap: false });
    expect(diff1).toMatchSnapshot();
    await orm.schema.execute(diff1.up);

    orm.discoverEntity(Book2, 'Book1');
    const diff2 = await orm.schema.getUpdateSchemaMigrationSQL({ wrap: false });
    expect(diff2).toMatchSnapshot();
    await orm.schema.execute(diff2.up);

    await expect(orm.schema.getUpdateSchemaMigrationSQL({ wrap: false })).resolves.toEqual({
      down: '',
      up: '',
    });
  });

});
