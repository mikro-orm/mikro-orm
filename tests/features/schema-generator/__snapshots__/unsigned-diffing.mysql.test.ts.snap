// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unsigned diffing in mysql schema generator updates column types when length changes (varchar, decimal, ...) 1`] = `
{
  "down": "alter table \`book\` modify \`price1\` int not null;
alter table \`book\` modify \`price2\` decimal(10,0) not null;
alter table \`book\` modify \`price3\` double not null;
",
  "up": "alter table \`book\` modify \`price1\` int unsigned not null;
alter table \`book\` modify \`price2\` numeric(10,0) unsigned not null;
alter table \`book\` modify \`price3\` double unsigned not null;
",
}
`;

exports[`unsigned diffing in mysql schema generator updates column types when length changes (varchar, decimal, ...) 2`] = `
{
  "down": "alter table \`book\` modify \`price1\` int unsigned not null;
alter table \`book\` modify \`price2\` decimal(10,0) unsigned not null;
alter table \`book\` modify \`price3\` double unsigned not null;
",
  "up": "alter table \`book\` modify \`price1\` int not null;
alter table \`book\` modify \`price2\` numeric(10,0) not null;
alter table \`book\` modify \`price3\` double not null;
",
}
`;
